### PR TITLE
Updated TextField component to implement AutoPropertyDiffable

### PR DESCRIPTION
**Pull request by aducret**
This PR adds AutoPropertyDiffable logic to TextField component.

## Screencast
pr

### CODE REVIEW

File: Portal/View/Components/TextField.swift [hunk 1]
```diff
@@ -57,6 +57,14 @@ public struct TextFieldEvents<MessageType> {
         )
     }
     
+    public func toArray() -> [(UIControlEvents, MessageType?)] {
```
Comment (guidomb) [inline Portal/View/Components/TextField.swift]: I don't think this needs to be public

File: Portal/View/UIKit/Renderers/Utils/MessageHandler.swift [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        
+        let mailbox: Mailbox<MessageType> = getMailbox(mailboxKey: &mailboxAssociationKey)
+        let dispatcher = MessageDispatcher(mailbox: mailbox, message: message)
+        self.register(dispatcher: dispatcher, for: event, dispatcherKey: &messageDispatcherAssociationKey)
+        self.addTarget(dispatcher, action: dispatcher.selector, for: event)
+        
+        return mailbox
+    }
+    
+    func getDispatcher<MessageType>(for event: UIControlEvents,
```
Comment (guidomb) [inline Portal/View/UIKit/Renderers/Utils/MessageHandler.swift]: Why is it necessary to have a `dispatcherKey` parameter? I cannot think of a case where an instance of `UIControl` would have more than one dispatcher

File: Portal/View/UIKit/Renderers/Utils/MessageHandler.swift [hunk 3]
```diff
@@ -0,0 +1,90 @@
+//
+//  MessageHandler.swift
+//  Portal
+//
+//  Created by Argentino Ducret on 8/23/17.
+//  Copyright © 2017 Guido Marucci Blas. All rights reserved.
+//
+
+protocol MessageHandler where Self : UIControl {
```
Comment (guidomb) [inline Portal/View/UIKit/Renderers/Utils/MessageHandler.swift]: I like what you did with this protocol, my suggestion would be to drop the `dispatcherKey` in `getDispatcher` and also avoid the `mailboxKey` in `getMailbox`

File: Portal/View/UIKit/Renderers/TextFieldRenderer.swift [hunk 4]
```diff
@@ -18,50 +20,109 @@ internal struct TextFieldRenderer<MessageType, RouteType: Route>: UIKitRenderer
...
     
+    // TODO: Delete
     fileprivate func apply(style: TextFieldStyleSheet) {
         let size = CGFloat(style.textSize)
         style.textFont.uiFont(withSize: size)     |> { self.font = $0 }
         style.textColor                           |> { self.textColor = $0.asUIColor }
         style.textAligment                        |> { self.textAlignment = $0.asNSTextAligment }
     }
     
+    fileprivate func getMailbox<MessageType>() -> Mailbox<MessageType> {
```
Comment (guidomb) [inline Portal/View/UIKit/Renderers/TextFieldRenderer.swift]: Why do you have this method? `UITextField` already conforms to `MessageHandler`

File: Portal.xcodeproj/project.pbxproj [hunk 5]
```diff
@@ -315,6 +317,14 @@
 /* End PBXFrameworksBuildPhase section */
 
 /* Begin PBXGroup section */
+		46952C311F4DB07F0075D536 /* Utils */ = {
```
Comment (guidomb) [inline Portal.xcodeproj/project.pbxproj]: I would remove `Utils` directory and put `MessageHandler` inside `Renderers` directory

File: Portal/View/Components/TextField.swift (hunk 1 above)
Comment (guidomb) [reply Portal/View/Components/TextField.swift]: Also this method should not be here, this is `UIKit` specific and should be added as an extension inside `TextFieldRederer.swift` file.

File: Portal/View/UIKit/Renderers/TextFieldRenderer.swift [hunk 6]
```diff
@@ -18,50 +20,109 @@ internal struct TextFieldRenderer<MessageType, RouteType: Route>: UIKitRenderer
...
+                self.font = font.uiFont(withSize: fontSize)
+                
+            case .textSize(let textSize):
+                let fontName = self.font?.fontName
+                fontName |> { self.font = UIFont(name: $0, size: CGFloat(textSize)) }
+            }
+        }
+    }
     
+    // TODO: Delete
```
Comment (guidomb) [inline Portal/View/UIKit/Renderers/TextFieldRenderer.swift]: Why didn't you delete this?

File: Portal/View/UIKit/Renderers/TextFieldRenderer.swift [hunk 7]
```diff
@@ -18,50 +20,109 @@ internal struct TextFieldRenderer<MessageType, RouteType: Route>: UIKitRenderer
...
+                apply(events: events)
             }
-            properties.onEvents.onEditingEnd |> { _ = textField.dispatch(
-                message: $0,
-                for: .editingDidEnd,
-                with: mailbox)
+        }
+    }
+    
+    fileprivate func apply<MessageType>(events: TextFieldEvents<MessageType>) {
```
Comment (guidomb) [inline Portal/View/UIKit/Renderers/TextFieldRenderer.swift]: I would refactor this as
```swift
fileprivate func apply<MessageType>(events: TextFieldEvents<MessageType>) {
        for (event, maybeMessage) in events.messageByEvent {
            if let message = maybeMessage {
                _ = self.on(
                    event: event, dispatch: message,
                    dispatcherKey: &messageDispatcherAssociationKey, mailboxKey: &mailboxAssociationKey)
            } else {
                _ = self.unregisterDispatcher(
                    for: event,
                    dispatcherKey: &messageDispatcherAssociationKey) as MessageDispatcher<MessageType>?
            }
        }
}
```

File: Portal/View/Components/TextField.swift (hunk 1 above)
Comment (guidomb) [reply Portal/View/Components/TextField.swift]: I would make this a computed property named `messageByEvent`

**Review by guidomb — CHANGES_REQUESTED** [review]

**Comment by aducret** [comment]
[Hacer que TextField sea diffeable en Portal](https://trello.com/c/OblWjP5K/291-hacer-que-textfield-sea-diffeable-en-portal)

File: Portal/View/UIKit/Renderers/TextFieldRenderer.swift [hunk 8]
```diff
@@ -18,63 +20,84 @@ internal struct TextFieldRenderer<MessageType, RouteType: Route>: UIKitRenderer
...
+                self.font = font.uiFont(withSize: fontSize)
+                
+            case .textSize(let textSize):
+                let fontName = self.font?.fontName
+                fontName |> { self.font = UIFont(name: $0, size: CGFloat(textSize)) }
+            }
+        }
+    }
+    
+    fileprivate func getMessagesByEvent<MessageType>(events: TextFieldEvents<MessageType>) -> [(UIControlEvents, MessageType?)] {
```
Comment (guidomb) [inline Portal/View/UIKit/Renderers/TextFieldRenderer.swift]: Just a small detail, but what I was suggesting before was to add an extension to `TextFieldEvents` inside `TextFieldRenderer.swift` file.
```swift
fileprivate extension TextFieldEvents {

  fileprivate var messagesByEvent: [(UIControlEvents, MessageType?)] {
    return [
            (.editingDidBegin, events.onEditingBegin),
            (.editingChanged, events.onEditingChanged),
            (.editingDidEnd, events.onEditingEnd)
        ]
  }

}
```

**Review by guidomb — CHANGES_REQUESTED** [review]
@aducret Remember to add tests for the `TextFieldRenderer`

File: PortalTests/TextFieldRenderSpec.swift [hunk 9]
```diff
@@ -0,0 +1,239 @@
...
+class TextFieldRenderSpec: QuickSpec {
+    override func spec() {
+        
+        var layoutEngine: LayoutEngine!
+        
+        beforeEach {
+            layoutEngine = YogaLayoutEngine()
+        }
+        
+        describe(".apply(changeSet: ButtonChangeSet) -> Result") {
```
Comment (guidomb) [inline PortalTests/TextFieldRenderSpec.swift]: Rename this. It should be `".apply(changeSet: TextFieldChangeSet) -> Result"`

File: PortalTests/TextFieldRenderSpec.swift [hunk 10]
```diff
@@ -0,0 +1,239 @@
...
+                    textField.textAligment = .center
+                }
+                
+                changeSet = TextFieldChangeSet.fullChangeSet(
+                    properties: textFieldProperties,
+                    styleSheet: textFieldStyle,
+                    layout: layout())
+            }
+            
+            context("when the change set contains button property changes") {
```
Comment (guidomb) [inline PortalTests/TextFieldRenderSpec.swift]: fix copy paste

**Review by guidomb — CHANGES_REQUESTED** [review]

**Pull request merged**